Add a Share a Draft feature

brandonpayton opened the pull request:
# Share a Draft

This is a WIP PR to add a Share a Draft feature to Calypso so users may request feedback on drafts prior to publishing. This is the main PR for the feature UI in the post settings sidebar. We may need smaller, supporting PRs, including a separate PR for the share-related nag banner in the editor.

A live branch with these changes is here:
https://calypso.live/?branch=add/share-a-draft

The plan is to prototype with a mock data layer, share with design, and iterate.

I'm maintaining the following checklists as a personal reminder and to reflect general progress.

**NOTE:** We are discussing first shipping a simpler version 1 without feedback comments, so I've broken the lists into version 1 and version 2 categories. There are some items in version 2 which are desirable for version 1, but I am placing them with version 2 to increase the chance we ship something sooner rather than later.

## Version 1 - Without feedback comments

### Development

These are necessary items for a simpler version 1 without comments.

- [x] Create invitation to open feedback view
- [x] Create form for requesting feedback
- [x] Create component to view a share with comments
- [x] Create unit-tested action types and creators
- [x] Update `keyedReducer` to support explicit key removal -- supporting share removal
- [x] Create unit-tested reducer for draft feedback state
- [x] Create unit-tested selector to list draft feedback for a given site and post
- [x] Integrate existing UI with redux
- [x] Add UI for revoke/restore share access
- [x] Validate email input
- [ ] Update Gravatar component to support mutually exclusive `email` and `url` props instead of a `user` prop
- [ ] Add query component for requesting feedback for a draft
- [ ] Add necessary actions and data-layer support for connecting this with a real REST API
- [ ] Error reporting via `createNotice` in data-layer
- [x] Document initial vision for the REST API and include it in a README with the draft feedback data-layer provider
- [ ] Component unit tests
- [ ] Refine styles
- [ ] Review accessibility -- the settings sidebar doesn't appear accessible to me, but I need to confirm with the team.
- [ ] Review RTL support
- [ ] Review `translate` calls and improve communication for translators
- [ ] Consider usage stats
- [ ] Address and remove remaining inline TODOs

### Bugs

- [ ] The settings sidebar generates setState errors when navigating to feedback view before settings view fully loads
- [ ] The feedback view disappears when resized to narrow window dimensions (and probably for many mobile devices in portrait orientation)

## Version 2 - With comments

### Development

- [ ] Support submitting more than one email address at a time
- [ ] Add confirmation dialog for revoke share access
- [ ] Create or update Accordion component that can be driven via props rather than state -- this depends on whether UX continues to require only one section to be expanded at a time.
- [ ] Show collapsed accordion-like header when feedback-request-form is collapsed so the user knows how to expand it.
- [ ] Timestamp feedback comments
- [ ] Highlight unread feedback in the feedback sidebar


## Questions for design

- [ ] How should the UI change after a post is published? The feedback UI is launched through an invitation to request feedback, but the invitation no longer makes sense after publishing.
- [ ] Should a share be able to be expanded when there are no comments, and if so, what should be shown?


cc @nb @dmsnell

brandonpayton commented:
I had been working on the dialog present in the early mockup but discarded it when the design change. Now that this PR exists, that kind of work will have more visibility.

nb commented on client/post-editor/editor-sidebar/index.jsx:
```diff
@@ -27,6 +28,11 @@ export default class EditorSidebar extends Component {
 		isPrivate: PropTypes.bool,
 	}
 
+	_viewFeedbackPane = () => {
```
What is the purpose of the leading underscore?

brandonpayton replied:
It is there because of an old habit of marking members intended to be private and because I saw it [used](https://github.com/Automattic/wp-calypso/blob/f97ae1a398aa1f5ad210877b7a58a68dc2ca384e/client/components/dialog/dialog-base.jsx#L42) in the Calypso codebase. But I'm glad you asked because I'm not sure it is useful when used with React. It's not common that we directly call methods on other components.

I plan to remove the underscore and avoid the practice in other components, unless someone knows a good reason we should continue.

nb commented:
Thanks, @brandonpayton, this is an awesome start – the component will help us see the feature in a perspective from the beginning.

Something that can be helpful is adding a couple of screenshots, so that passers by don't need to run the code.

Also, for the same passers by: https://calypso.live/?branch=add/share-a-draft

a8ci18n commented on client/post-editor/editor-sidebar/feedback-header.jsx:
```diff
@@ -0,0 +1,38 @@
+				compact borderless
+				className="editor-sidebar__back"
+				onClick={ closeFeedback }
+				title={ translate( 'Close feedback' ) }
```
Hi! I've found a possible matching string that has already been translated **99** times:
`translate( 'Feedback' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

dmsnell replied:
While it seems pretty far-fetched that this could be taken to mean, "Feedback which is nearby," since it _is_ a translation string and it _is_ ambiguous, we can add a `context` param to let the translators know this is a verb

```js
translate( 'Close feedback', { context: 'verb: imperative' } )
```

Context means that there are multiple ways this could be used in a sentence depending on context and we are describing how to clear it up. Sometimes a string is just a little awkward and needs a little extra nudge to the translators: in these cases we also have `comment`. Each context creates a new translation string but comments only tag along to provide hints, so they are preferred when the string isn't context-sensitive.

```js
translate( 'Approved', { comment: 'indicates thing has been approved blarg blurp' } )
```

See also for `Share Draft` (is it a verb or a noun?)

brandonpayton replied:
Thank you! It's good to know and will help me embrace our role in the translation effort.

I'll review the existing strings and update as needed.

brandonpayton commented:
I began adding a semi-styled skeleton of the feedback sidebar UI.

The sidebar currently uses local UI `state` to decide whether to display the settings or the feedback view, but this needs to be moved to the redux store because we'd like to reset it in response to sidebar actions and because we want to be able to trigger the feedback view from outside the sidebar (via the feedback nag banner in the editor).

The next steps are to flesh out the share list, mock feedback in the data layer, and wire up redux state and actions. Two interesting bits of UI for this are the Gravatar icons and the accordion-like behavior of the feedback list.

The Gravatar component takes a `user` prop that is a user object, but we need to be able to show Gravatars for people who aren't users. I'm considering a PR that extends the component to support an email address string, but I don't like the idea of an overloaded interface. In addition, we may want to show the user avatar for users and fallback to the email-based gravatar for non-users, requiring us to query for a user. This is based on my understanding that a Gravatar user may specify a WordPress-specific avatar that is used instead.

The Calypso Accordion component is a single header with a collapsible body. Multiple Accordion components can be placed next to one another for a complete accordion effect, but I believe the sections operate independently so that opening one does not close the others. But the feedback list is designed with the full accordion behavior where opening one item closes the others. My current plan is simply to implement a custom component with this behavior and to generalize later if desired.

Here are some screenshots of the current progress.

Settings Sidebar:
![settings sidebar](https://user-images.githubusercontent.com/530877/28008162-3bd7b074-650b-11e7-901d-f146fd57c4f9.png)

Feedback Sidebar:
![feedback sidebar](https://user-images.githubusercontent.com/530877/28008167-406d388e-650b-11e7-9905-81a596bb6479.png)

dmsnell commented:
> The Gravatar component takes a user prop that is a user object, but we need to be able to show Gravatars for people who aren't users…In addition, we may want to show the avatar for WordPress users (requiring us to query for a user) and fall back to an email-based gravatar for non-users.

I'm pretty sure we don't expose other users' email addresses on account of the privacy violation it could raise. If we connected the email address to a WordPress.com account that could open that door up. I would think that simply passing an email address to the component would be sufficient.

In fact, I'm curious why `<Gravatar />` even takes a user object since it's inherently tied to an email and not a user. Maybe @v18 could help bring in some context. In any case, I think a change in API could help out. We appear to have around 20-30 uses of `<Gravatar />` and several of those pass nothing (to get a placeholder) so such a move could be done without a huge PR.

```js
<Gravatar email="boigo@example.com" alt="Boidlefram Gotyuba" />

<Gravatar email={ this.props.user.avatar_URL } alt={ this.props.user.display_name } />
```

v18 commented:
> I'm considering a PR that extends the component to support an email address string, but I don't like the idea of an overloaded interface

I agree with @dmsnell, Gravatar doesn't need to take a user object, and can instead receive the separate properties the user object provides. Currently they are:

- `display_name`
- `avatar_URL`
- `ID` - needed for displaying a temp image when a user has just finished uploading a Gravatar in Calypso, should be able to be be null without breaking things

All of the above can either be omitted or have good defaults for a non-registered user.

> I'm curious why <Gravatar /> even takes a user object

I don't have any more context for this, from what I recall while working with the code, using the user object was just easier to pass in at first iteration.

dmsnell commented:
> ID - needed for displaying a temp image when a user has just finished uploading a Gravatar in Calypso, should be able to be be null without breaking things

@v18 this doesn't seem inherent to the `<Gravatar />` component. it seems more like we could handle this on the parent component or make a HOC wrapper for `<WordPressAvatar />` that takes the user object/injects the temporary URL. Should only need the temporary for the logged-in user right? So I think the number of times we need the temporary URL are small and probably best contained in the parent's logic.

v18 commented:
@dmsnell yes, the temp URL is needed very rarely.

dmsnell commented:
> @dmsnell yes, the temp URL is needed very rarely.

okay @v18 that makes sense, but I was just a bit hasty when I should have emphasized that I don't think that relates to the user at all. that is, instead of passing in a user ID, we should just pass in a temporary image. even better yet is if that temporary image always loads so that there is something immediately available while the real one comes in. I'm thinking of a default inline SVG or something.

v18 commented:
@dmsnell yeh, we could pass in the temporary image URL, but as you mention that would have to be done using a parent, connected component, since the temp URL is received from the redux store.

> even better yet is if that temporary image always loads so that there is something immediately available while the real one comes in

The temp image URL is probably badly named, but it's only relevant to users who have recently uploaded a new Gravatar using [`<EditGravatar />`](https://github.com/Automattic/wp-calypso/tree/master/client/blocks/edit-gravatar). It's not referring to a temporary image that gets displayed before the real one loads.

There isn't a universal fallback image right now, either for when it hasn't finished loading the image or when the download fails. The Gravatar in the masterbar has an [SVG](https://github.com/Automattic/wp-calypso/blob/8db70cf3a584cd679ed884f25337da3b2140a397/client/layout/masterbar/logged-in.jsx#L109) which acts as a fallback, but the rest of the Gravatars don't.

brandonpayton commented:
If the component is parameterized with the image URL, should it just be called `<Avatar />` since there would be nothing tying it to Gravatar but the assumption that URLs would include a Gravatar host like "https://www.gravatar.com/avatar/<email-hash>"?

brandonpayton commented:
Also, thanks for all your feedback on this.

dmsnell commented:
@v18 how is it a problem to move that into the parent? It sounds like the need for the URL overwrite is an exception rather than a norm

v18 commented:
@dmsnell not a problem, was re-iterating what you said 👍

brandonpayton commented:
I was wrong in my previous statement about the component having nothing tying it to Gravatar. I remembered that it does add a size-related query param when requesting from a gravatar domain.

dmsnell commented on client/post-editor/editor-feedback-invitation/index.jsx:
```diff
@@ -0,0 +1,42 @@
+			</Button>
+		</div>
+	);
+}
```
Fun fact! (and purely personal style) we can remove some syntax noise here with the arrow function syntax. Further, we can export the component itself _unwrapped_ so as to make testing it in isolation easier.

```js
export const EditorFeedbackInvitation = ( { onTrigger, translate } ) => (
	<div …>
		…
	</div>
);

export default localize( EditorFeedbackInvitation );

// other.jsx
import EditorFeedbackInvitation from …

<EditorFeedbackInvitation onTrigger={ grillABurger } />

// test.jsx
import 

const invite = shallow( <EditorFeedbackInvitation translate={ identity } onTrigger={ triggerSpy } /> );
expect( invite.blah.blah ).to.equal( 'Send to a Friend' );
```

brandonpayton replied:
I like the arrow version in your example but prefer creating named functions where possible. It's probably not a big deal in practice, especially for disciplined React code, but after debugging a lot of code with many layers of indirection and anonymous functions, it makes me feel good to create a function that will have a name in the debugger's stack UI. :)

Thanks for the solid suggestion of exporting the unwrapped component to facilitate testing. I'll update that now.

dmsnell replied:
That's fine. Note too however that even the arrow-function syntax we will have a `displayName` on the React component. I'm fine with using the `function` syntax I just prefer not to. Also are you sure if won't get a name in debugging?

```js
const Test = () => {
	throw new Error( 'hi' );
	return <div>Test</div>;
}
```

<img width="421" alt="screen shot 2017-07-12 at 1 03 26 pm" src="https://user-images.githubusercontent.com/5431237/28114763-83969a30-6702-11e7-9826-a82bee99706c.png">

brandonpayton replied:
Interesting! I know I've been frustrated by the issue with anonymous functions and assumed it would be an issue with arrow functions as well. I'm guessing that the devtools have improved to address this pain, which is cool. Thanks for showing me this.

I no longer have a reason to avoid arrow functions for components, except that I like reading `function SomeHandyComponent() {}`. :)

dmsnell replied:
Well `function` still has a few advantages, namely that it's statically available whereas `const` expressions need to be defined before they are used. I wonder too at the way the engine handles them differently too but my own style leads me to prefer the arrow functions and "wait and see" if there are bigger issues.

It's true that the dev tools have improved to support the patterns in use which is a good thing!

dmsnell commented on client/post-editor/editor-sidebar/feedback-header.jsx:
```diff
@@ -0,0 +1,38 @@
+FeedbackSidebarHeader.propTypes = {
+	translate: PropTypes.func,
+	closeFeedback: PropTypes.func,
+};
```
I'm sure you are planning on adding them, but above you have default props but not here in this component. Also above you had `translate` as required but here optional.

brandonpayton replied:
Honestly, it was an oversight as I was working to get _something_ out there we could iterate on. I would self-review later but could miss it. Thanks for calling this out.

I'm updating this PR's components to clearly and consistently mark required properties.

dmsnell commented on client/post-editor/editor-sidebar/feedback-list.jsx:
```diff
@@ -0,0 +1,62 @@
+import Accordion from 'components/accordion';
+import Gravatar from 'components/gravatar';
+
+class FeedbackList extends PureComponent {
```
we've had a good number of discussions on this and I don't know where we ended up with `Component` vs `PureComponent`. `react-redux` already makes a component pure by default so I tend to discourage mixing both pure bases (using `Component` instead) but since we're not wrapped with the Redux helper I suppose this here is appropriate.

brandonpayton replied:
Since I'll be wrapping with the Redux helper as a next step, I'll make a note to switch to `Component` in those cases. Does that work for you?

dmsnell commented on client/post-editor/editor-sidebar/feedback-list.jsx:
```diff
@@ -0,0 +1,62 @@
+			<div className="editor-sidebar__feedback-list">
+				{ sharedLinks.map(
+					sharedLink => this.renderSharedLink( sharedLink )
+				) }
```
Fun fact! There's no need to create a new closure here since `map` will already inject the `sharedLink` parameter…

```js
{ sharedLinks.map( this.renderSharedLink ) }
```

brandonpayton replied:
The purpose of the arrow function was to bind to the context since the rendering helper required it.

I like the visibility and decomposition arguments for not splitting out sub-render functions. I'm fixing the components to respect this now.

dmsnell commented on client/post-editor/editor-sidebar/feedback-list.jsx:
```diff
@@ -0,0 +1,62 @@
+						? translate( 'No feedback yet.' )
+						: this.renderComments( comments ) }
+				</Accordion>
+			</div>
```
do we need this `<div>` here or is there a way to get the CSS class applied directly to the accordion? or a way to select the accordion without the `<div>`? It just looks a bit funny with a single component wrapped in a `<div>`

brandonpayton replied:
I currently don't have a need for that container or class, so I'm removing it. Thanks!

dmsnell commented on client/post-editor/editor-sidebar/feedback-list.jsx:
```diff
@@ -0,0 +1,62 @@
+				{ comments.map(
+					// NOTE: It should be OK to use the index for `key` because
+					// the list is currently append-only
+					( comment, index ) => ( <li key={ index }>{ comment }</li> )
```
Thanks for the comment, but why not just use the `comment` itself since we know that should be safe without a disclaimer? It may not always be append-only and this may not get updated.

brandonpayton replied:
This may be a pathological example, but comments could be identical. What do you think?

dmsnell replied:
It's a good point but I bet low risk. If we're afraid of it we could also add a timestamp or some other attribute from the comment data. The point was just that it seemed easy to use a real key, so why not? Maybe it's not easy…

dmsnell commented on client/post-editor/editor-sidebar/feedback-request-form.jsx:
```diff
@@ -0,0 +1,36 @@
+			<div className="editor-sidebar__feedback-request-form">
+				<p>{ description }</p>
+				<label>
+					{ translate( 'Friend\'s Email' ) }
```
feel free to use double quotes or backticks here to avoid escaping; some have different opinions on what should happen here

brandonpayton replied:
Good to know. Done.

dmsnell commented on client/post-editor/editor-sidebar/feedback-view.jsx:
```diff
@@ -0,0 +1,84 @@
+	constructor() {
+		super();
+		this.state = { openFeedbackCount: 0 };
+	}
```
Fun fact! We can avoid creating the constructor function when internal state doesn't depend on `props`

```js
static propTypes = { derp };

state = { openFeedbackCount: 0 }
```

brandonpayton replied:
I saw you'd enabled the class properties transform but hadn't thought of using it here. It reads more clearly to me, and I made the change.

dmsnell commented on client/post-editor/editor-sidebar/feedback-view.jsx:
```diff
@@ -0,0 +1,84 @@
+
+	isAllFeedbackClosed() {
+		return this.state.openFeedbackCount === 0;
+	}
```
often I like to try and pull all of the methods that I can out of the React component and make them functions in the module scope. granted, we probably won't have thousands of this component instantiated but it's an easy way to trim down the size of the component in memory and reduce GC churn. Also it makes these helpers easy to test.

```js
export const haveShares = ( { sharedLinks } ) => sharedLinks.length > 0;
export const isAllFeedbackClosed = ( { openFeedbackCount } ) => openFeedbackCount == 0;

// later

haveShares( this.props );
isAllFeedbackClosed( this.state );
```

of course, sometimes when we do this we recognize some patterns we could reuse…

```js
export const haveShares = links => links.length > 0;
export const isAllFeedbackClosed = count => count === 0;

// later

haveShares( this.props.sharedLinks );
isAllFeedbackClosed( this.state.openFeedbackCount );
```

then it can seem silly to have so much code extracted for such simple operations

```js
const { sharedLinks } = this.props;
const { openFeedbackCount } = this.state;

return (
	…
	{ openFeedbackCount === 0 && this.renderRequestForm() }
	{ sharedLinks.length && this.renderFeedbackList() }
	…
);
```

So in the end we may decide that the helper functions weren't all that much better at communicating what we're doing, though with `this.renderXYZ()` some of that meaning and coupling between the condition and the render is lost, whereas directly embedding them might complete the circle.

```js
	{ openFeedbackCount === 0 && (
		<div>
			<div className="…" />
			<FeedbackRequestForm />
		</div>
	) }
	{ sharedLinks.length && (
		<div…>
			…
		</div>
	) }
```

brandonpayton replied:
I agree with you about extracting functions and then also about the functions not really making things clearer. I'll drop the helpers.

Also, it's cool to see your example with `( booleanExpression && ( reactElement ) )`. I wasn't sure whether React would render "false" for a `false` value, but my testing says that it doesn't, which is cool. I did discover that `( zero && ( reactElement ) )` results in a zero being rendered, so the expressions need to evaluate to real boolean values, not just falsey values.

dmsnell commented on client/post-editor/editor-sidebar/feedback-view.jsx:
```diff
@@ -0,0 +1,84 @@
+					{ translate( 'Friends' ) }
+				</div>
+				<FeedbackList sharedLinks={ sharedLinks }
+					onToggleFeedback={ this.onToggleFeedback } />
```
Nitpick: we tend to prefer splitting these on their own lines and then leaving the trailing `/>` on a newline. Feel free to use `prettier` to format your code. We have a fork included in Calypso so you can tell your IDE to use it or just run `./node_modules/.bin/prettier --write XYZ.js`

brandonpayton replied:
Thanks for letting me know. I set up format-on-save with the prettier fork and think its a big improvement. And I like having to worry less about subtle style decisions.

dmsnell commented on client/post-editor/editor-sidebar/index.jsx:
```diff
@@ -27,8 +29,31 @@ export default class EditorSidebar extends Component {
+	};
+	closeFeedbackPane = () => {
+		this.setState( { showFeedback: false } );
+	};
```
Fun fact! We're only evaluating an expression so we can get rid of the noisy curlies!

```js
openFeedbackPane = () => this.setState( { showFeedback: true } )
```

brandonpayton replied:
I have mixed feelings about this. On one hand, I prefer expressions to imperative statements, but on the other, I don't like returning the value of an expression unless it is explicitly intended. In some cases, the latter can be a problem depending on whether the caller optionally consumes a return value. In this case, I doubt setState returns anything, and React doesn't appear to use the return value of event listeners to do things like preventDefault.

All of that said... I think your sample reads better and will change my code to match.

Thanks

dmsnell replied:
yeah here we know the return value isn't being used, but also we have to acknowledge that _every_ function in JavaScript returns a value, so even if we use the function body syntax we're still returning `undefined`, which ironically, is a value.

brandonpayton replied:
The main idea is to be conscious not to accidentally return something of meaning to the caller. The only concrete one I can think of is the DOM0 event handler example where `anchorElement.onclick = () => functionThatReturnsFalse()` prevents the default action when the link is clicked. It's usually not an issue, but I believe I've caused at least one bug in other situations by preferring the expression form of arrow function and incidentally returning a value.

This is now probably way too much discussion for a minor point. :) I like the expression form here and will update the lines.

dmsnell commented:
This is definitely looking nice @brandonpayton, and _props_ to you for tossing in the ARIA props. I recommend seeing if we couldn't merge some of the render functions as that's typically our style: to prefer larger `render()` functions and fewer `renderThis()` or `renderThat()` functions. I also left some translation notes that would be good to think about; translation is a subject matter of its own…

Anyway, feel free to consider the styling prompts and make a choice on your own.

brandonpayton commented:
Thanks for the review @dmsnell ! I made updates for most of your comments but still need to review translation for opportunities to add clarity.

a8ci18n commented on client/post-editor/editor-sidebar/feedback-header.jsx:
```diff
@@ -0,0 +1,40 @@
+				borderless
+				className="editor-sidebar__back"
+				onClick={ closeFeedback }
+				title={ translate( 'Close feedback' ) }
```
Hi! I've found a possible matching string that has already been translated **99** times:
`translate( 'Feedback' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **99** times:
`translate( 'Feedback' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

dmsnell commented on client/post-editor/editor-sidebar/index.jsx:
```diff
@@ -10,7 +10,26 @@ import EditorDrawer from 'post-editor/editor-drawer';
+		],
+	},
+	{
+		label: 'brandon@happycode.net',
```
May not be the optimum choice to expose real email addresses here in the test data…

brandonpayton replied:
Wow, reused an address from a mockup and didn't think twice about making it public. I feel terrible and will correct it immediately.

brandonpayton replied:
I edited the history to use fake email addresses and force-pushed the changes, so the latest commits referenced by the PR are OK. Thank you for pointing this out.

dmsnell replied:
Long live GitHub comments!

brandonpayton replied:
Ha, yeah. They are unfortunately long-lived and tied to the previous commits. I've learned neither address owner is concerned at this point, which is good.

brandonpayton commented:
The UI is still missing some things, but I'd like to get what is there connected to redux. Unfortunately, it looks like editor state is split across flux store(s) and redux. I'm looking at what is present on the redux side to find the right place to maintain the state for draft shares. If there is no clear home for this state, I plan to create a slice at the top level, and we can move later if desired.

brandonpayton commented:
I spent tonight reviewing component data and action dependencies and examining how state is queried, organized, and updated within Calypso. I feel a bit more comfortable but have a long way to go to understanding the big picture. There is a wonderfully large amount of state.

For state, it seems like we will need:
1. A new property under editor state to drive whether the feedback view should be shown.
2. A new state slice for draft feedback, containing a site->post->shares structure.

According to the Calypso approach, I believe I need to define a query component for loading a drafts' shares and feedback. Is that correct, @nb?

I didn't want to end the night without tangible output, so I added the minimum number of action types necessary for design to play with the UI. Next, I plan to use these to maintain a `draftFeedback` state slice without respect to site or post so we can begin driving the UI with dynamic data. Afterward, we can refine the slice to consider site and post.

dmsnell commented:
> If there is no clear home for this state, I plan to create a slice at the top level, and we can move later if desired.

Seems like a good plan. Redux works best with flat trees. Make sure you check out the `keyedReducer()` which might be able to help you focus on the items themselves and organize the relationships to sites and posts.

brandonpayton commented:
Thanks for pointing me to `keyedReducer`, @dmsnell . It's elegant. I looked for the author because I wanted to thank them as well, and I found out it was you. :)

Tonight, I added action creators and tests and created the reducer but am not getting the output I expect in the reducer tests. I pushed the work I have so far, including the failing tests, which is why CircleCI is currently showing failure.

I'll debug and push an update to the reducer as soon as I'm able.

brandonpayton commented:
The reducer's issue has to do with how I was adding a placeholder object for feedback sharing when starting to edit a post. `keyedReducer` noticed that the placeholder was equivalent with the initial state and was not adding the placeholder under the keys as I expected.

This is probably the right behavior for `keyedReducer`, and I'm not sure adding a placeholder object was valuable to begin with.

If I can get more time today, I'll fix the reducer and finish the tests.

dmsnell commented on client/state/draft-feedback/reducer.js:
```diff
@@ -0,0 +1,71 @@
+				return state;
+		}
+	},
+);
```
Nice work on this state stuff @brandonpayton. I would like to suggest however two bits here: use idiomatic Redux composition; don't create functions in the argument list of a function call (something I like to avoid in general unless for very simple arrow-functions)

We have `combineReducers()` which is a good fit for this state structure. I prefer to try and _never_ have state which is an object, instead relying on composition of reducers over simple data types instead.

The second bit is that with this use of `keyedReducer()` we can't test the inner reducer because they too coupled together. My habit is to create something like a `_Item` reducer and then fold over the `keyedReducer()`. This means a couple of things: first of all, we get a separate testable reducer, then second of all we can use the reducer outside of the wrapper.

```js
export const isEnabled = ( state = true, { type } ) => {
	switch ( type ) {
		case DRAFT_FEEDBACK_SHARE_REVOKE:
			return false;
		case DRAFT_FEEDBACK_SHARE_RESTORE:
			return true;
		default:
			return state;
	}
}

export const comments = ( state = [], { comments, type } ) =>
	type === DRAFT_FEEDBACK_COMMENTS_ADD
		? [ ...state, ...comments ]
		: state;

export const draftShareItem = combineReducers( {
	comments,
	isEnabled,
} );

export const draftShare = keyedReducer( 'emailAddress', draftShareItem );
```

here also I have tried to more clearly indicate the boolean type of the `enabled` since it could be just as easy to imagine `enabled` to mean a list of email addresses for which the share is enabled etc…

dmsnell commented on client/state/draft-feedback/reducer.js:
```diff
@@ -0,0 +1,71 @@
+			}
+		} ),
+	),
+);
```
It's confusing, but _without persistence_ is now the default, so you shouldn't need this wrapper.

dmsnell replied:
I see what you are doing here, but I wonder if we might be better off with some wrapper around the keyed reducers instead of inside of them. we have had discussions about gating the keyed reducers and acting on more broad action types (for example, one which should affect every item in the collection) and our conclusion at the time was that it makes most sense to do this from a composed wrapper instead of embedding the logic.

what would that look like here?

well I think that this is confusing because in some cases it looks like we are using a reducer and in others not. what if instead we passed the draft share reducer into the other…

```js
export const withEditor = reducer => ( state, action ) => {
	switch ( action.type ) {
		case EDITOR_START:
			return {};

		case EDITOR_STOP:
			return undefined;

		default:
			reducer( state, action );
	}
}

export const draftShares = keyedReducer( 'siteId', keyedReducer( 'postId', withEditor( draftShare ) ) );
```

I don't know how much better this is than yours and I still wonder if it should work at a higher level, but it's an idea. You can think about it and make your own decisions.

brandonpayton commented:
I fixed the reducer issue I mentioned above but will be making further updates in response to code review comments.

The need to explicitly remove keys led to PR #16275 for updating `keyedReducer` to remove keys when their value is being set to `undefined`.

I am continuing focus on connecting the UI to actions and dynamic data.

dmsnell commented on client/post-editor/editor-sidebar/feedback-comments.jsx:
```diff
@@ -0,0 +1,25 @@
+				( comment, index ) =>
+					<li key={ index }>
+						{ comment }
+					</li>,
```
is the comma here a typo?

brandonpayton replied:
I thought so, but it turns out it is inserted by `prettier`.

dmsnell commented on client/post-editor/editor-sidebar/feedback-view.jsx:
```diff
@@ -0,0 +1,64 @@
+		const { openFeedbackCount } = this.state;
+
+		this.setState( {
+			openFeedbackCount: openFeedbackCount + ( isOpen ? 1 : -1 ),
```
this might need a comment explaining what is going on

brandonpayton replied:
I thought the code could be made clearer, so I attempted to do that. But I did end up adding a comment on the initial state assignment. What do you think of the updates?

dmsnell commented on client/post-editor/editor-sidebar/index.jsx:
```diff
@@ -10,7 +10,26 @@ import EditorDrawer from 'post-editor/editor-drawer';
+		],
+	},
+	{
+		label: 'email2@share.test',
```
fun fact! any `@example.com` email address is by-spec not routable, meaning you can safely use them without worrying that someone might by happenstance actually have that email address.

brandonpayton replied:
Thank you for mentioning that. I'd update them if I wasn't working to remove them right now.

samouri commented:
> The reducer's issue has to do with how I was adding a placeholder object for feedback sharing when starting to edit a post. keyedReducer noticed that the placeholder was equivalent with the initial state and was not adding the placeholder under the keys as I expected.


> The need to explicitly remove keys led to PR #16275 for updating keyedReducer to remove keys when their value is being set to undefined.

You may be interested in the discussion [here](https://github.com/Automattic/wp-calypso/pull/14245).  

I actually added a way to remove a key, but its not the most explicit.
The gist is that when a subreducer is created dynamically, its fair to treat its non-existence as the same thing as it being set to its initialState.  Therefore we made it so that when keyedReducer receives the `initialState` for a key, it deletes the key.

Whats your use-case for needing to remove a key?  Would this method make sense for you?

brandonpayton commented:
Hi @samouri, thanks for linking to that discussion. It's good to see the thought behind the implementation.

My use case for removing a key is to correct for optimistic updates where the update was not successful. In this case, sharing a draft with an email address adds a key for that email address under `draftFeedback[siteId][postId][emailAddress]`, but if the request to share is not successful, the key should be removed.

I first wrote a version that handled the remove action by returning the initial state, but it depended on knowledge of `keyedReducer` implementation and wasn't clear to read. Explicitly returning `undefined` was clearer to read.

It was this:
```javascript
case REMOVE:
    return initialState;
```
Versus this:
```javascript
case REMOVE:
    return undefined;
```

What do you think?

samouri commented:
> My use case for removing a key is to correct for optimistic updates where the update was not successful. In this case, sharing a draft with an email address adds a key for that email address under draftFeedback[siteId][postId][emailAddress], but if the request to share is not successful, the key should be removed.

What is expected to happen in the ui when there is failure? 
- Do we want to remove the list item + dispatch an error notice?  ( --> remove key from keyedReducer).
- Or do we want to keep it in the ui but visually marked as a failure somehow ( --> keeping the key within the collection).

brandonpayton commented:
> What is expected to happen in the ui when there is failure?

That is a good question. I'm not sure we know at this point. I'm sketching the state shape and transitions and getting this to a place where design try it out and we can iterate. Later, it may turn out that we don't want to remove a key (or at least not immediately). But I'm thinking we've at least encountered a valid use case for key removal with `keyedReducer`.

brandonpayton commented:
I updated the reducer in response to review comments and added a `get-draft-feedback-shares` selector. I think there may be a few review comments from yesterday that I haven't addressed, so I'm going to review those first and then connect the UI with redux.

brandonpayton commented:
I connected the existing UI to redux, and it is now possible to add requests for feedback to the store.

I'd like to start sharing with design but plan to add a temporary way for them to add comments to shares. Currently, it is only possible via manually dispatching DRAFT_FEEDBACK_COMMENT_ADD actions. Once the temporary measure is in place, I plan to continue fleshing out and refining the UI and adjusting state shape and actions as necessary.

I need to review and revise the checklist for this PR, but that will have to wait for tomorrow.

dmsnell commented:
> it is only possible via manually dispatching DRAFT_FEEDBACK_COMMENT_ADD actions. Once the temporary measure is in place, I plan to continue fleshing out and refining the UI and adjusting state shape and actions as necessary.

that's fine. just leave the instructions with an example dispatch in the *Testing* section of the PR description :wink:

brandonpayton commented:
I wanted to leave instructions for dispatching comment-add actions but didn't think there was an easy way for them to obtain the site ID which is required for the action. Do you have any suggestions, @dmsnell?

brandonpayton commented:
Nevermind. I discovered that `getState` is available during dev as well and can offer simple instructions for this. Thanks for the suggestion!

dmsnell commented:
> to obtain the site ID

also this is not hard but you found a good solution with state.

brandonpayton commented:
> What is expected to happen in the ui when there is failure? 
> - Do we want to remove the list item + dispatch an error notice?  ( --> remove key from keyedReducer).
> - Or do we want to keep it in the ui but visually marked as a failure somehow ( --> keep the key in keyedReducer).

Hi @samouri, I revisted your questions about error handling and was surprised by my original reply since I knew at least part of the answer. I appreciate your feedback and am sorry for not giving you a better answer.  Let's try this again.

I envision that we will optimistically add a share and then remove the share in the event of an error. This is the reason I wanted to update `keyedReducer` to support more explicit removal. The part I was not sure about is how to best reflect errors in the state. I can imagine adding an `error` property to `state.draftFeedback` to reflect errors at the post level and then adding an `error` property at the share level for errors revoking and restoring access to the share (something our user can do). It is less clear to me when the error state should be reset. Perhaps, it should be reset after another action successfully updates the state.

Is there a particular approach you favor? Do you have any other thoughts or suggestions?

brandonpayton commented:
Tonight, I took some time to clean up styles and component structure. Tomorrow, I need to take time to update the PR overview with what is left for the admin side of this feature. After that, I plan to add the revoke/restore buttons and dialogs and implement email validation.

dmsnell commented:
Concerning errors, do we need to store and display that state or might it be enough to announce it to the customer via `createNotice()`?

brandonpayton commented:
Thanks for pointing me to `createNotice()`, @dmsnell. As a user, I'd prefer seeing that kind of notification outside of the feedback UI. I'll confirm this approach with the design team.

brandonpayton commented:
This evening, I added revoking and restoring access.

![revoke-restore-access](https://user-images.githubusercontent.com/530877/28489835-47fd6c34-6e81-11e7-8cbc-7e232753c139.png)

nb commented:
Thanks for getting the prototype in our hands so quickly, @brandonpayton 🚤 I will test it a bit and leave any feedback either here or at the design discussion.

nb commented:
Currently sharing with an empty e-mail seems to work:

![empty-address](https://user-images.githubusercontent.com/27954/28493747-78cdcc34-6f24-11e7-9956-3fd6a7cfa8b4.gif)

brandonpayton commented:
> Currently sharing with an empty e-mail seems to work:

Yes, we currently share with even the non-existent. :) Email validation is coming soon.

brandonpayton commented:
I added docs under `client/data-layer/draft-feedback` describing a vision for this feature's REST API.
* [`REST-API-V1.md`](https://github.com/Automattic/wp-calypso/blob/f90fb75a05bcc049999ccce84a9735d7a9f92737/client/state/data-layer/wpcom/draft-feedback/REST-API-V1.md) describes an API for an initial simpler version of this feature.
* [`REST-API-V2.md`](https://github.com/Automattic/wp-calypso/blob/f90fb75a05bcc049999ccce84a9735d7a9f92737/client/state/data-layer/wpcom/draft-feedback/REST-API-V2.md) describes an API for the current design, including comments and highlighting unread feedback.

One thing that is missing is who is authorized to use which API endpoints. In general, it seems like permission should be limited to those who are authorized to edit a given post. The one exception to this is the endpoint for adding a comment where all that is needed is the ID of the feedback request.

As always, any feedback is appreciated.

dmsnell commented:
Nice API docs @brandonpayton though you should be aware that we only do GET and POST operations 😢 - it's a long story and hopefully one day will change.

❓ why is the email address optional?

brandonpayton commented:
Thanks, @dmsnell. That's unfortunate about the HTTP methods. Do you know if you ever support PATCH-like functionality via POST?

The email address is optional in anticipation of supporting anonymous feedback links, which we support in wp-admin's request feedback feature.

dmsnell commented:
> Do you know if we ever support PATCH-like functionality via POST?

It's really up to the endpoint developer 🙃 but we typically don't do it.

two things though:
 - one approach I have taken in the past is simply to post the new item and let the server perform a diff on that object to grab the changes: `POST /sites/{ siteId }/posts/{ postId }/shares-whatever/{ shareId }/`
 - we have a way to lie using something like `_method: 'PATCH'` in the `POST` body. you might scour some docs or in `wpcom.js` for that as I think our server entry-point can rewrite those methods as if they actually were the intended HTTP method.

brandonpayton commented:
It's good to have your thoughts on this. Thanks, @dmsnell.

For now, I think I will leave the API docs as-is so the person implementing the API can make what they wish of the general intention. Then we can update the docs.

I have another question that touches on REST API design. The Share a Draft form typically allows multiple email addresses to be entered at once, but each address ideally maps to a single `feedback-request` resource. According to the HTTP spec, I believe this means we should make a separate POST request for each, but it is not uncommon to compromise and support a single request that speaks to multiple resources. Do you know offhand how Calypso and the WordPress REST API tend to approach these kinds of cases? If not, I can look into it further but wanted to ask.

dmsnell commented:
> Do you know offhand how Calypso and the WordPress REST API tend to approach these kinds of cases?

Off-hand I can't think of too many resources that we create in batches. I don't think we need to be worrying about optimizing that away anyhow.

nb commented:
@dmsnell it changed long time ago, especially for new resources  :) Everything new uses the WP-API framework, which is classic REST. Look at v2 APIs [in the console](https://developer.wordpress.com/docs/api/console/).

samouri commented on client/post-editor/editor-sidebar/feedback-share.jsx:
```diff
@@ -0,0 +1,81 @@
+							// NOTE: It should be OK to use the index for `key` because
+							// the list is currently append-only
+							( comment, index ) =>
+								<p key={ index } className="editor-sidebar__feedback-share-comment">
```
do comments have any other identifier we can use? Can you use the text of the comment as the key?

using the index is really just getting around the react warning without addressing the issue.

brandonpayton replied:
Hi @samouri, @dmsnell mentioned the same thing, and I agree it is less than ideal. I'd rather not use a comment as a key because comments can technically be identical while indices will not, and IIUC, the worst that can happen if our comment list is reordered or loses a comment is sub-optimal patching of a small list of simple elements. Still, I don't like this.

What do you think of using a hash of each comment? If we took this approach, it may be better to maintain the hash with the data model than to hash with each render.

We could consider using the timestamp which needs to be added to the comment structure anyway, but I don't think we can guarantee uniqueness.

Another option is to rely on the server to simply tag each comment with a UUID.

What do you think?

brandonpayton replied:
I should also add that I know there can be more problematic issues for improperly keyed stateful components, which makes me less happy with the compromise. It's better to avoid cheating or having to explain and remember a special case.

dmsnell replied:
let's just leave it in. a hash is overkill. You might change the comment to something more aggressive though:

```js
// using the list index is best here because comment text _can_ be identical across comments
// it's breaking a rule, but here it's appropriate
```

samouri replied:
I read the React docs on the issue -- you are totally right @brandonpayton.  Its not as big a deal as I thought -- i'm totally okay with leaving as-is 👍

brandonpayton replied:
Thanks, guys. I made the update.

samouri commented on client/post-editor/editor-sidebar/feedback-view.jsx:
```diff
@@ -0,0 +1,116 @@
+
+		return { siteId, postId, shares };
+	},
+	dispatch =>
```
Theres a really nice shorthand that lets you skip `bindActionCreators`. Instead of supplying a function like 

```js
dispatch => bindActionCreators( { creator1, creator1 } )
```

You can just supply an object like this:
```js
{ creator1, creator1 }
```

So in your case it would look something like:
```js
connect(
  mapStateToProps,
  { addDraftShare, revokeDraftShare, restoreDraftShare }
)( localize( FeedbackView ) );
```

[source](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options)

brandonpayton replied:
Thank you for sharing this.

samouri commented on client/state/draft-feedback/reducer.js:
```diff
@@ -0,0 +1,92 @@
+export const emails = ( state = [], { type, emailAddress } ) => {
+	switch ( type ) {
+		case DRAFT_FEEDBACK_SHARE_ADD:
+			return state.indexOf( emailAddress ) >= 0 ? state : state.concat( emailAddress );
```
what happens if someone types in the same email address multiple times?

brandonpayton commented:
I added email validation to `feedback-request-form` and think the implementation is reasonable. But I'm concerned that it might be better to use `client/lib/form-state` and related components for the sake of consistency (e.g., consistent validation debounce times). I didn't use`form-state` initially because I didn't know about it and because it looked more complicated than I wanted for a single-field form component. Any thoughts on this?

It may be that I just need to learn how to use it.

I also have concerns about the accessibility of the form. The validation error message is not currently tied to the email input. There appears to be friction using attributes like `aria-describedby` because those attributes require the use of DOM IDs, and AFAIK, React doesn't naturally generate unique DOM IDs for components' elements. I see [one Calypso component](https://github.com/Automattic/wp-calypso/blob/6f4a5ee5741beaa1e000bc2d0bf0d5a737df7bd9/client/components/select-dropdown/index.jsx#L271) that works around this by generating a component-specific `instanceId`, but I think we should find an approach that requires less work at a component level. If each component gets a unique ID, IDs for child elements can be generated like `` `${id}-email-input` ``. Do you have any thoughts on this? Do we have a solution that I am missing?

brandonpayton commented:
I'm marking this as on hold since we are pursuing a simpler design first. It's likely the design will be reworked enough that this PR is not directly applicable, but I'm not going to close it yet.

brandonpayton commented:
I'm closing this in favor of #16611.

brandonpayton commented:
Thanks for all the feedback while we pursued this design.
